bftengine/src/bcstatetransfer/SourceSelector.cpp

```diff
@@ -175,7 +187,49 @@ void SourceSelector::selectSource(uint64_t currTimeMilli) {
   fetchingTimeStamp_ = 0;
   fetchRetransmissionCounter_ = 0;
   receivedValidBlockFromSrc_ = false;
+
+  LOG_INFO(logger_, "Selected new source replica " << currentReplica_);
+}
+
+void SourceSelector::updateCurrentPrimary(uint16_t newPrimary) {
+  if (currentPrimary_ == newPrimary) return;
+  ConcordAssert(isValidSourceId(newPrimary));
```

Comment:
Rethinking about this one,  might this be dangerous?
Should we assert? It is an external value, the source id is usually secured and validated by the communication module so at first it looks fine. But-  ST is unaware of that - who knows where it was fetched from in the future. If it is fetched from the message body it might be byzantine. Let's just replace it with an if + error + handle an error if happened.
If there was an error - I would reset all the counters for the nominated and exit - since it means something is very "unstable" and we want to replace primary when things are "stable".